src/java/org/apache/cassandra/db/marshal/UserType.java

```diff
@@ -425,28 +403,15 @@ public <V> boolean referencesUserType(V name, ValueAccessor<V> accessor)
     @Override
     public UserType withUpdatedUserType(UserType udt)
     {
-        if (!referencesUserType(udt.name))
-            return this;
+        // If we're not the UDT to update, we can rely on the default implementation
+        if (!name.equals(udt.name))
+            return (UserType) super.withUpdatedUserType(udt);
 
-        // preserve frozen/non-frozen status of the updated UDT
-        if (name.equals(udt.name))
-        {
-            return isMultiCell == udt.isMultiCell
-                 ? udt
-                 : new UserType(udt.keyspace, name, udt.fieldNames(), udt.fieldTypes(), isMultiCell);
-        }
-
-        return new UserType(udt.keyspace,
-                            name,
-                            fieldNames,
-                            Lists.newArrayList(transform(fieldTypes(), t -> t.withUpdatedUserType(udt))),
-                            isMultiCell());
-    }
-
-    @Override
-    public boolean referencesDuration()
-    {
-        return fieldTypes().stream().anyMatch(f -> f.referencesDuration());
+        // TODO reinvestiate this method
+        assert udt.isMultiCell;
```

Comment:
I think we mentioned this on Slack. The updated UDT passed as and argument is a schema definition, and those are always non-frozen.